ws/config.py

```diff
@@ -1,180 +1,232 @@
-    if not os.path.isdir(string):
-        raise configargparse.ArgumentTypeError("directory '%s' does not exist" % string)
-    return string
-
-# list of comma-separated items from a fixed set
-def argtype_comma_list_choices(choices):
-    choices = set(choices)
-
-    def wrapped(string):
-        items = [item.strip() for item in string.split(",")]
-        for item in items:
-            if item not in choices:
-                raise configargparse.ArgumentTypeError("unknown item: '{}' (available choices: {})".format(item, choices))
-        return items
-
-    return wrapped
-
-def getArgParser(subname=None, *args, **kwargs):
-    if subname is None:
-        _, _script = os.path.split(sys.argv[0])
-        subname, _ = os.path.splitext(_script)
-
-    # create config file parser
-    cfp = ConfigFileParser("site", subname)
-    kwargs["config_file_parser"] = cfp
-    kwargs["ignore_unknown_config_file_keys"] = True
-
-    # set brief usage parameter by default
-    kwargs.setdefault("usage", "%(prog)s [options]")
-    # TODO: the supplied description is merged with the one from configargparse, which looks very ugly, regardless of the used formatter
-    kwargs.setdefault("formatter_class", configargparse.RawDescriptionHelpFormatter)
-
-    ap = configargparse.ArgParser(*args, **kwargs)
-
-    # add config file and site arguments
-    ap.add_argument("-c", "--config", metavar="PATH", is_config_file_arg=True,
-            help="path to config file (default: %(default)s)")
-    ap.add_argument("--site",
-            help="sets the top-level section to be read from config files (default: %(default)s)")
+        raise argparse.ArgumentTypeError(
+            "cannot convert '{}'to boolean".format(string))
+
+
+def argtype_configfile(string):
+    """Convert `--config` argument to an absolute file path."""
+    dirname = os.path.dirname(string)
+    name, ext = os.path.splitext(os.path.basename(string))
+
+    # Configuration name was specified.
+    if not dirname and not ext:
+        config_dir = os.getenv('XDG_CONFIG_HOME',
+                               os.path.expanduser(CONFIG_DIR))
+        path = os.path.join(config_dir,
+                            "{}/{}.conf".format(PROJECT_NAME, string))
+    # Relational or absolute path was specified.
+    elif:
+        if ext != '.conf':
+            raise argparse.ArgumentTypeError(
+                "config filename must end with '.conf' suffix")
+        path = os.path.abspath(os.path.expanduser(string))
+
+    return path
+        
+def argtype_path(string):
+    """Convert a string to an absolute filesystem path."""
+    return os.path.abspath(os.path.expanduser(string))
+
+
+# ==============
+# Action classes
+# ==============
+
+class _ActionCheckPath(argparse.Action):
+
+    def __call__(self, parser, namespace, path, option_string=None):
+        if not os.path.lexists(path)
+            raise FileNotFoundError(
+                "no such file or directory: {}".format(path)
```

Comment:
I don't think this should be checked with `action="check_path"`. For example, you cannot combine it with `nargs="+"` or `action="append"` or `action="extend"`.

Reply:
Actually, not every built-in action can be combined with this params (e.g., `"store_true"`). `"check_path"` and `"check_dirname"` must be used on a single argument representing a filesystem path. Do you have any other use cases in mind?

Reply:
Ok, let's talk about this request. In the old-old days, there was two argtypes: `argtype_existing_dir` and `argtype_dirname_must_exist`. Both of them deal with the strings representing the filesystem path, and do the checks: the first checks if the whole path string represents the existing dir, and the second checks only the 'dirname' part of the path. I have found the opportunity to split the functionality. Now there is uniform type `argtype_path` and two actions, `'check_path'` (which now can check not only the directories, but files and links too), and `'check_dirname'`.

There is no big problem to add `nargs` support to actions code as well as some list-working functionality like `'append'` and `'extend'` do. But I need clear instructions on what kind of program behavior you expect.

And yes, I suddenly realized where I got this `os.path.abspath(os.path.expanduser(string))` construct. I took it from you. :) These two argtypes works on it. Is there any special purpose of using `abspath` or it can be wiped out?

Reply:
So after "splitting the functionality", you have 3 objects instead of 2 to take care of. You also have limited functionality, or would have to add non-trivial amount of the same code to `check_path` and `check_dirname` to support `nargs` and `action="append"` and `action="extend"`. This is not a good direction...

Consider how you'd implement an argument taking integer values: you'd use `type=int`, because you want to convert `str` to `int` objects. And see, the `int()` "function" also does two things: it deals with the representation of integers and it checks if the argument is an integer. How would you implement an argument taking a positive integer? You would most likely create a custom function which converts the argument to `int` (and thus checks that the argument is an integer) and then checks if the value is positive. Or you could create a class such as `PositiveInt` which inherits from `int` and does the check in the `__init__` method. These two approaches are almost the same, but IMO the latter is over-complicated.

The `argtype_existing_dir` and `argtype_dirname_must_exist` functions did the same - they created some representation of a path and did some checks. The fact that the path representation uses `str` as the underlying type is irrelevant (since Python 3.4 it could use some object from [pathlib](https://docs.python.org/3/library/pathlib.html), but that is for another discussion). The command to get the path representation is so simple that there is no refactoring needed (if it was too complicated, there would be another `argtype_path` function which would just return the representation and the other argtype functions would call it before doing the checks). I really think this is a better approach than the `check_*` actions, and many other people likely think so too: the [first thing I could find on the web](https://stackoverflow.com/questions/38834378/path-to-a-directory-as-argparse-argument) uses `type` as well...

Reply:
> So after "splitting the functionality", you have 3 objects instead of 2 to take care of.

You forgot about `argtype_config` which must be checked too. Action `"check_path"` can do this check, and your argtypes cannot. So it will be 3 against 3 at minimum. If one day you decide to create, for example, `argtype_file` (not for config), it also could be checked by `"check_path"`. My approach is more flexible and extensible, do you see?

I add nargs support for actions:  60acdfa 

About pathlib - I had investigated this module some days ago. I had found that the only one really usefult feature can be adopted from it - the class constructor `pathlib.PosixPath`, which can be passed to `type=` directly. All the path checking functionality must be coded by yourself anyway.

Reply:
I did not even get to `argtype_config` yet, you can add the check just before its `return` statement to get rid of `check_path`. It's unrelated to the arguments above...